rioxarray/raster_index.py

```diff
@@ -0,0 +1,428 @@
+import textwrap
+from collections.abc import Hashable, Mapping
+from typing import Any
+
+import numpy as np
+import pandas as pd
+from affine import Affine
+from xarray import DataArray, Index, Variable
+from xarray.core.coordinate_transform import CoordinateTransform
+
+# TODO: import from public API once it is available
+from xarray.core.indexes import CoordinateTransformIndex, PandasIndex
+from xarray.core.indexing import IndexSelResult, merge_sel_results
+
+
+class AffineTransform(CoordinateTransform):
```

Comment:
I think the `Affine*` classes should live in Xarray. They are useful for both geo and bio-imaging domains at least (and possibly astro too).

Reply:
Yes I agree these are useful in many domains! The wrapped `affine.Affine` object is here restricted to 2D affine transformation of the plane so it might be limited for general usage, though. These classes are also thin wrappers.

Reply:
They are "thin" in code but 'deep' in understanding of the Xarray model. That's why I think we should vendor it. Even so, vendoring will save a bunch of your time as you iterate on this since you wouldn't need to keep code in two places in-sync.

Reply:
Maybe the whole content of `raster_index.py` including the `RasterIndex` class could (eventually) be moved into Xarray or some dedicated package? There's nothing IO specific and it could be useful in other downstream packages like odc-geo, geoxarray, etc.